### PR TITLE
Include file statistics domain in IcebergSplit

### PR DESCRIPTION
**Pull request by raunaqmorarka**
## Description
Allows pruning splits on workers using late arriving dynamic filters without opening file format footers
Also helps in avoiding unnecessary predicate pushdown into ORC/parquet

## Additional context and related issues
Matches delta lake connector approach of including `statisticsPredicate` in DeltaLakeSplit

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java [hunk 1]
```diff
@@ -194,17 +202,14 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
...
             this.pushedDownDynamicFilterPredicate = dynamicFilter.getCurrentPredicate()
                     .transformKeys(IcebergColumnHandle.class::cast)
                     .filter((columnHandle, domain) -> isConvertableToIcebergExpression(domain));
             TupleDomain<IcebergColumnHandle> fullPredicate = tableHandle.getUnenforcedPredicate()
                     .intersect(pushedDownDynamicFilterPredicate);
             // TODO: (https://github.com/trinodb/trino/issues/9743): Consider removing TupleDomain#simplify
             TupleDomain<IcebergColumnHandle> simplifiedPredicate = fullPredicate.simplify(ICEBERG_DOMAIN_COMPACTION_THRESHOLD);
-            boolean usedSimplifiedPredicate = !simplifiedPredicate.equals(fullPredicate);
-            if (usedSimplifiedPredicate) {
+            if (!simplifiedPredicate.equals(fullPredicate)) {
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:212]: nit: I find the comment below confusing:
```
// Pushed down predicate was simplified, always evaluate it against individual splits
```
Do we assume that Iceberg is slow if we didn't simplify predicate?
Do we want to evaluate it per split because of partition columns or data columns pruning?
Why do we set `pushedDownDynamicFilterPredicate` to `all`? (I think it means we actually don't want to push simplified version of it)
Why not always evaluate it against splits and never push?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java [hunk 2]
```diff
@@ -249,7 +255,8 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
                 FileScanTask wholeFileTask = fileScanIterator.next();
                 boolean fileHasNoDeletions = wholeFileTask.deletes().isEmpty();
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:257]: nit: We could extract `dynamicFilterPredicate` before:
```
            this.pushedDownDynamicFilterPredicate = dynamicFilter.getCurrentPredicate()
                    .transformKeys(IcebergColumnHandle.class::cast)
                    .filter((columnHandle, domain) -> isConvertableToIcebergExpression(domain));
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 3]
```diff
@@ -418,26 +420,51 @@ else if (identity.getId() == TRINO_MERGE_PARTITION_DATA) {
                 deletePredicate);
     }
 
-    private TupleDomain<IcebergColumnHandle> getEffectivePredicate(
+    private TupleDomain<IcebergColumnHandle> getUnenforcedPredicate(
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:423]: nit: restructure could be separate commit

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergNodeLocalDynamicSplitPruning.java [hunk 4]
```diff
@@ -147,6 +147,7 @@ public void testDynamicSplitPruningOnUnpartitionedTable()
                     PartitionData.toJson(new PartitionData(new Object[] {})),
                     ImmutableList.of(),
                     SplitWeight.standard(),
+                    TupleDomain.all(),
                     ImmutableMap.of());
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergNodeLocalDynamicSplitPruning.java:151]: Add a unit test for your changes in `TestIcebergNodeLocalDynamicSplitPruning`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 5]
```diff
@@ -448,7 +461,9 @@ private TupleDomain<IcebergColumnHandle> prunePredicate(
         }
         // Filter out partition columns domains from the dynamic filter because they should be irrelevant at data file level
         return unenforcedPredicate
-                .filter((columnHandle, domain) -> !partitionKeys.containsKey(columnHandle.getId()));
+                .filter((columnHandle, domain) -> !partitionKeys.containsKey(columnHandle.getId()))
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:464]: Move the comment 

```
// Filter out partition columns domains from the dynamic filter because they should be irrelevant at data file level
```

above this line

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 6]
```diff
@@ -418,25 +417,37 @@ else if (identity.getId() == TRINO_MERGE_PARTITION_DATA) {
...
-    private TupleDomain<IcebergColumnHandle> getEffectivePredicate(
+    private TupleDomain<IcebergColumnHandle> getUnenforcedPredicate(
             Schema tableSchema,
             Map<Integer, Optional<String>> partitionKeys,
-            TupleDomain<IcebergColumnHandle> dynamicFilterPredicate,
+            DynamicFilter dynamicFilter,
             TupleDomain<IcebergColumnHandle> unenforcedPredicate)
     {
-        TupleDomain<IcebergColumnHandle> effectivePredicate = unenforcedPredicate.intersect(dynamicFilterPredicate);
-        if (dynamicFilterPredicate.isAll() || dynamicFilterPredicate.isNone() || partitionKeys.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:428]: I've moved that change into next commit now to make it clearer that it's needed to let the `remove domains from predicate that fully contain split data` code execute

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTypes.java [hunk 7]
```diff
@@ -165,7 +165,7 @@ public static Object convertIcebergValueToTrino(Type icebergType, Object value)
         if (icebergType instanceof Types.StringType) {
             // Partition values are passed as String, but min/max values are passed as a CharBuffer
             if (value instanceof CharBuffer) {
-                value = new String(((CharBuffer) value).array());
+                value = ((CharBuffer) value).toString();
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTypes.java:168]: It is an existing bug that I discovered through test failures from my changes
